Backport of Avoid blocking child type updates on parent ack into release/1.13.x

hc-github-team-consul-core opened the pull request:
## Backport

This PR is auto-generated from #15083 to be assessed for backporting due to the inclusion of the label backport/1.13.


**WARNING** automatic cherry-pick of commits failed. Commits will require human attention.

> merge conflict error: POST https://api.github.com/repos/hashicorp/consul/merges: 409 Merge conflict []



The below text is copied from the body of the original PR.

---

### Description
When providing updates to Envoy it is safest to follow the update order of: Clusters, Endpoints, Listeners, Routes. This order is encoded in our delta xDS response generation code.

Another noteworthy attribute of these resources is that there are parent->child relationships between: Cluster->Endpoints and Listener->Routes. Envoy couples the storage of a parent and a child type such that whenever a control plane sends an update for a parent type, it MUST also send fresh data for the child type, even if Envoy previously had the latest data for the child type.

A wrinkle that isn't accounted for currently is that Envoy does not actually ACK parent types until after attempting to receive data for the child types. This behavior also actually differs between Clusters and Listeners:
- When a cluster is sent, Envoy will wait up to a 15s timeout for endpoints to arrive before sending the cluster ACK.
- When a listener is sent, and it specifies RDS routes, Envoy will wait until those routes arrive before ACKing the listener. Though it's not clear to me what the timeout is for this, if any, since it exceeds a minute from my tests.

However, a behavior encoded in our xDS update order is that we avoid sending ANY data if we are waiting on ACKs for ANY resource. Meaning that when we first send a cluster to Envoy and it requests endpoints for that cluster, the endpoints do not actually get sent for at least 15s. This is because the endpoint update is paused by Consul until the cluster ACK, which Envoy pauses until it gets endpoints or times out.

This commit changes the xDS order gating such that we only block listener/route updates if there are cluster/endpoin updates pending. This is to avoid routing to an invalid destination.

### Testing & Reproduction steps
* Ran manual tests for various scenarios against a live Envoy to validate assumptions (Envoy v1.21 and v1.23)
* Updated the unit tests to more accurately reflect Envoy's behavior
* Anecdotally the integration tests seem to be running ~1min faster than usual with this patch

### Links
Likely related to: https://github.com/envoyproxy/envoy/pull/5887

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [x] not a security concern


---

<details>
<summary> Overview of commits </summary>

  - e43e898c212ede4237213d4000c8bad6ea8e5144  - 3d0b984cc8e8bca98aff0fa5f74f4de1227427c9  - c47a0c394753412cfcce07b37ecf48753524029d  - 5581010b7996690c2fe4235042208430552c7c29  - dea2bd91c1be814c4f13d039edb067c9944bfa67 

</details>

Commit:
no-op commit due to failed cherry-picking

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/consul?pullRequest=15517)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/consul?pullRequest=15517)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**1** out of **2** committers have signed the CLA.
- [x] analogue
- [ ] temp

----

**temp** seems not to be a GitHub user.
You need a GitHub account to be able to sign the CLA.
If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).


<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/consul?pullRequest=15517).</sub>

analogue committed:
Avoid blocking child type updates on parent ack

When providing updates to Envoy it is safest to follow the update order
of: Clusters, Endpoints, Listeners, Routes. This order is encoded in our
delta xDS response generation code.

Another noteworthy attribute of these resources is that there are
parent->child relationships between: Cluster->Endpoints and
Listener->Routes. Envoy couples the storage of a parent and a child type
such that whenever a control plane sends an update for a parent type, it
MUST also send fresh data for the child type, even if Envoy previously
had the latest data for the child type.

A wrinkle that isn't accounted for currently is that Envoy does not
actually ACK parent types until after attempting to receive data for the
child types. This behavior also actually differs between Clusters and
Listeners:
- When a cluster is sent, Envoy will wait up to a 15s timeout for
  endpoints to arrive before sending the cluster ACK.
- When a listener is sent, and it specifies RDS routes, Envoy will wait
  until those routes arrive before ACKing the listener. Though it's not
  clear to me what the timeout is for this, it exceeds a minute from my
  tests.

However, a behavior encoded in our xDS update order is that we avoid
sending ANY data if we are waiting on ACKs for ANY resource. Meaning
that when we first send a cluster to Envoy and it requests endpoints for
that cluster, the endpoints do not actually get sent for at least 15.
This is because the endpoint update is paused by Consul until the
cluster ACK, which Envoy pauses until it gets endpoints or times out.

This commit changes the xDS order gating such that we only block
listener/route updates if there are cluster/endpoin updates pending.
This is to avoid routing to an invalid destination.

analogue committed:
Fixup flaky test
